File: src/CImplementations/stackProblems/testSortStack.c
#ifndef TESTSORTSTACK

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <assert.h>
#include "sortStack.c"

void testSort(){

	Node* stack = newStack();
	assert(stack == NULL);
	sortStackAsc(&stack);
	assert(stack == NULL);
}


void testSortI(){

	Node* stack = newStack();
	assert(stack == NULL);	
	push(&stack,3);
	push(&stack,10);
	push(&stack,5);
	push(&stack,1);
	push(&stack,4);
	push(&stack,7);
	push(&stack,9);

	assert(peek(stack) == 9);

	Node* sortedStack = sortStackAsc(&stack);

	assert(pop(&sortedStack) == 1);
	assert(pop(&sortedStack) == 3);
	assert(pop(&sortedStack) == 4);
	assert(pop(&sortedStack) == 5);
	assert(pop(&sortedStack) == 7);
	assert(pop(&sortedStack) == 9);
	assert(pop(&sortedStack) == 10);

	assert(stack == NULL);	
	
}

void testSortII(){

	Node* stack = newStack();
	assert(stack == NULL);	
	push(&stack,7);
	push(&stack,6);
	push(&stack,5);
	push(&stack,4);
	push(&stack,3);
	push(&stack,2);
	push(&stack,1);	

	Node* sortedStack = sortStackAsc(&stack);

	assert(pop(&sortedStack) == 1);
	assert(pop(&sortedStack) == 2);
	assert(pop(&sortedStack) == 3);
	assert(pop(&sortedStack) == 4);
	assert(pop(&sortedStack) == 5);
	assert(pop(&sortedStack) == 6);
	assert(pop(&sortedStack) == 7);

	assert(stack == NULL);	
	
}


void testSortIII(){

	Node* stack = newStack();
	assert(stack == NULL);	
	push(&stack,1);
	push(&stack,2);
	push(&stack,3);
	push(&stack,4);
	push(&stack,5);
	push(&stack,6);
	push(&stack,7);	

	Node* sortedStack = sortStackAsc(&stack);

	assert(pop(&sortedStack) == 1);
	assert(pop(&sortedStack) == 2);
	assert(pop(&sortedStack) == 3);
	assert(pop(&sortedStack) == 4);
	assert(pop(&sortedStack) == 5);
	assert(pop(&sortedStack) == 6);
	assert(pop(&sortedStack) == 7);

	assert(stack == NULL);	
	
}
int main(void)
{

	testSort();
	testSortI();
	testSortII();
	testSortIII();

return 0;
}
#endif

File: src/Chapter3/TestSortStack.java
package Chapter3;

import static org.junit.Assert.assertTrue;
import org.junit.Test;

public class TestSortStack {

	@Test
	public void testSort(){
		Stack s = new Stack(3);
		
		s.push(5);
		s.push(2);
		s.push(8);
		s.push(1);
		s.push(7);
		
		SortStack ss = new SortStack(s);
		
		Stack sortedStack = ss.sortStackAsc();
		
		assertTrue(sortedStack.top.value == 1);
		assertTrue(sortedStack.top.prev.value == 2);
		assertTrue(sortedStack.top.prev.prev.value == 3);
		assertTrue(sortedStack.top.prev.prev.prev.value == 5);
		assertTrue(sortedStack.top.prev.prev.prev.prev.value == 7);
		assertTrue(sortedStack.top.prev.prev.prev.prev.prev.value == 8);
	}
	
	
	@Test
	public void testSortI(){
		Stack s = new Stack(3);
		
		s.push(5);
		s.push(2);
		s.push(8);
		s.push(1);
		s.push(7);
		s.push(7);
		
		SortStack ss = new SortStack(s);
		
		Stack sortedStack = ss.sortStackAsc();
		
		assertTrue(sortedStack.top.value == 1);
		assertTrue(sortedStack.top.prev.value == 2);
		assertTrue(sortedStack.top.prev.prev.value == 3);
		assertTrue(sortedStack.top.prev.prev.prev.value == 5);
		assertTrue(sortedStack.top.prev.prev.prev.prev.value == 7);
		assertTrue(sortedStack.top.prev.prev.prev.prev.prev.value == 7);
		assertTrue(sortedStack.top.prev.prev.prev.prev.prev.prev.value == 8);
	}
	
	@Test
	public void testSortII(){
		Stack s = new Stack(7);
		
		s.push(6);
		s.push(5);
		s.push(4);
		s.push(3);
		s.push(2);
		s.push(1);
		
		SortStack ss = new SortStack(s);
		
		Stack sortedStack = ss.sortStackAsc();
		
		assertTrue(sortedStack.top.value == 1);
		assertTrue(sortedStack.top.prev.value == 2);
		assertTrue(sortedStack.top.prev.prev.value == 3);
		assertTrue(sortedStack.top.prev.prev.prev.value == 4);
		assertTrue(sortedStack.top.prev.prev.prev.prev.value == 5);
		assertTrue(sortedStack.top.prev.prev.prev.prev.prev.value == 6);
		assertTrue(sortedStack.top.prev.prev.prev.prev.prev.prev.value == 7);
	}
	
	@Test
	public void testSortIII(){
		Stack s = new Stack(1);
		
		s.push(2);
		s.push(3);
		s.push(4);
		s.push(5);
		s.push(6);
		s.push(6);
		
		SortStack ss = new SortStack(s);
		
		Stack sortedStack = ss.sortStackAsc();
		
		assertTrue(sortedStack.top.value == 1);
		assertTrue(sortedStack.top.prev.value == 2);
		assertTrue(sortedStack.top.prev.prev.value == 3);
		assertTrue(sortedStack.top.prev.prev.prev.value == 4);
		assertTrue(sortedStack.top.prev.prev.prev.prev.value == 5);
		assertTrue(sortedStack.top.prev.prev.prev.prev.prev.value == 6);
		assertTrue(sortedStack.top.prev.prev.prev.prev.prev.prev.value == 6);
	}
}
